src/main/java/com/autotune/analyzer/adapters/MetricMetadataAdapter.java

```diff
@@ -0,0 +1,77 @@
+package com.autotune.analyzer.adapters;
+
+import com.autotune.analyzer.utils.AnalyzerConstants;
+import com.autotune.common.data.metrics.AcceleratorMetricMetadata;
+import com.autotune.common.data.metrics.MetricMetadata;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import com.autotune.utils.KruizeConstants;
+
+
+import java.io.IOException;
+
+/**
+ * This adapter actually specifies the GSON to identify the type of implementation of MetricMetadata
+ * to serialize or deserialize
+ */
+public class MetricMetadataAdapter extends TypeAdapter<MetricMetadata> {
+
+    @Override
+    public void write(JsonWriter out, MetricMetadata value) throws IOException {
+        if (value == null) {
+            out.nullValue();
+            return;
+        }
+        out.beginObject();
+        if (null != value.getType())
+            out.name(KruizeConstants.JSONKeys.TYPE).value(value.getType());
+
+        if (value instanceof AcceleratorMetricMetadata accel) {
```

Comment:
If I understand it correctly, for every new type, we would require modifications in this everytime ?

Reply:
Yes, for every implementation of `MetricMetadata` we need to add a modification. We should tell the adapter which fields to look for and populate in the implemented class object.